google/cloud/forseti/scanner/audit/external_project_access_rules_engine.py

```diff
@@ -0,0 +1,232 @@
+import itertools
+import re
+from google.cloud.forseti.common.util import logger
+from google.cloud.forseti.scanner.audit import errors as audit_errors
+from google.cloud.forseti.scanner.audit import base_rules_engine as bre
+from google.cloud.forseti.common.gcp_type import resource as resource_mod
+from google.cloud.forseti.common.gcp_type import resource_util
+
+
+
+LOGGER = logger.get_logger(__name__)
+
+class ExternalProjectAccessRulesEngine(bre.BaseRulesEngine):
+    """Rules engine for External Project Access."""
+
+    def __init__(self, rules_file_path, snapshot_timestamp=None):
+        """Initialize.
+
+        Args:
+            rules_file_path (str): file location of rules
+            snapshot_timestamp (str): snapshot timestamp. Defaults to None.
+                If set, this will be the snapshot timestamp
+                used in the engine.
+        """
+        super(ExternalProjectAccessRulesEngine,
+              self).__init__(rules_file_path=rules_file_path)
+        self.rule_book = None
+
+    def build_rule_book(self, global_configs):
+        """Build ExternalProjectAccess rule book from the rules definition file.
+
+        Args:
+            global_configs (dict): Inventory configurations.
+        """
+        self.rule_book = \
+            ExternalProjectAccessRuleBook(global_configs,
+                                          self._load_rule_definitions())
+
+    # TODO: The naming is confusing and needs to be fixed in all scanners.
+    def find_policy_violations(self, user_email,
+                               project_ancestry,
+                               force_rebuild=False):
+        """Determine whether CloudSQL acls violates rules.
+
+        Args:
+            user_email (str): The ID of the project
+            project_ancestry (list): List of ancestries which turn 
+                                     out to a list of resources.
+            force_rebuild (bool): Force the rebuild of the rule book
+
+        Returns:
+             generator: A generator of rule violations.
+        """
+        violations = itertools.chain()
+        if self.rule_book is None or force_rebuild:
+            self.build_rule_book()
+
+        # If the ancestor found in the rule is not in the project_ancestry
+        # ancestors, then this is a rule violation
+        rules_violated = [rule for ancestor, rule in \
+                          self.rule_book.resource_rules_map.iteritems() if \
+                          ancestor not in project_ancestry]
+
+        for rule in rules_violated:
+            violations = itertools.chain(
+                violations,
+                rule.policy_violation(user_email, project_ancestry))
+
+        return violations
+
+class ExternalProjectAccessRuleBook(bre.BaseRuleBook):
+    """The RuleBook for ExternalProjectAccess resources."""
+
+    # Class variable for matching the ancestor during rule validation
+    ancestor_pattern = re.compile(r'^organizations/\d+$|^folders/\d+$')
```

Comment:
Does this account for multiple folders?  Which can be 4 nested layers.

Reply:
ATM, this regex is testing the ancestor resource provided in the rules.yaml where only one ancestor per rule can be listed.